### PR TITLE
feat(mcdu): Title line of Progress page displays flight phase

### LINKED ISSUE #3958: MCDU: Progress Page should show current vertical flight phase
**Is your feature request related to a problem? Please describe.**

The Progress page shows ECON CRZ (along with the flight number) in the title regardless of the current flight phase.

**Describe the feature you'd like to see implemented**

The Progress page title should display the current vertical flight phase along with the flight number.

**References**

Prog page chapter of Honeywell FMS manual

**Additional context**

I’m working on this myself and will open a PR soon.

Discord username (if different from GitHub): NONAmr2433 #8777

### PR DESCRIPTION
**Pull request by Username23-23**
Fixes #3958 

## Summary of Changes

- [x] ECON CRZ in title line removed
- [x] Title of PROG page displays current vertical flight phase in green
- [x] Second line displays “CRZ” above cruise altitude rather than the flightPhase variable 
- [x] Correct spelling for all flight phases (DONE excluded)
- [x] Spacing (both title line and line directly below)

## Screenshots (if necessary)

**Screenshots of Preflight phase:**
**Before:**
Screenshot 2021-03-14 162837
**After:**
New screenshots to compare with references
actualnw504

## References

Honeywell FMS manual states that current vertical flight phase should be displayed on title line of Progress page.
See comments for screenshots of references.

IRL pilot Zezze21#0975 confirmed that TO is displayed during preflight and provided this youtube video for reference:
 https://youtu.be/D7HjVlvKd8o?t=148
## Additional context

There are no Honeywell references for the DONE flight phase (Thales displays DONE though), so it is left blank.

ECON appears in Econ thrust mode, which is not implemented yet, so this will need to be addressed in a later PR.
I updated the spacing on the  lines below the title to match references.

Discord username (if different from GitHub): NONAmr2433 #8777

## Testing instructions

- While doing a flight, check the progress page to check that that the title changes when transitioning flight phases (nothing is showed in the done phase)
- Also check that text alignment matches what is shown in the screenshots

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js:99]:
```suggestion
            [flightPhase + " " + flightNo],
```

Comment (Username23-23) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js]: Thanks for your review! I will commit these changes.

**Comment by MisterChocker** [comment]
So, I've looked at our honeywell manual and found something.
Following the manual step 1, it displays CLB, CZR, DES, before that there are one or two fields for e.g. ECON or EO (engine out)

image

We'll have to check against step 2, to confirm this is still valid and await feedback from our pilots.

I've seen also PREFLIGHT, CLIMB, CRUISE, DESCENT. That could've been Thales tho.

**Comment by Username23-23** [comment]
> So, I've looked at our honeywell manual and found something.
> Following the manual step 1, it displays TO, CLB, CZR, DES, APPR, before that there are one or two fields for e.g. ECON or EO (engine out)
> 
> image
> 
> Also the color seems to be dynamic, wasn't able to find any reference why this is as of now.
> 
> image
> image
> 
> I assume it reads PREF and DONE for Preflight and Done phase.
> 
> We'll have to check against step 2, to confirm this is still valid and await feedback from our pilots.
> 
> I've seen also PREFLIGHT, CLIMB, CRUISE, DESCENT. That could've been Thales tho.

Thanks for the tip on EO.
As for PREF, what makes you think that? Most of the pictures I’ve seen show PREFLIGHT, but that could be Thales as you said.
DONE, I don’t have any references on, so I’ll try to look for some.

**Comment by MisterChocker** [comment]
@Username23-23 I just found the EO, I am not saying this inside the scope of this pr, I just thought pointing it out might be a good idea.

**Comment by Username23-23** [comment]
> @Username23-23 I just found the EO, I am not saying this inside the scope of this pr, I just thought pointing it out might be a good idea.

Alright, I’ve changed the summary address that. So that leaves EO and Econ to be in a separate PR.

**Review by beheh — APPROVED** [review]

**Comment by Username23-23** [comment]
> Just noticed the alignment of the text does not match the references. e.g.
> the first letter/number of the flight number always aligns two spaces left of rec max, or in other words the 3 number or letter of the flight number aligns with the R of REC MAX.
> 
> Additionally the flight phase is has always one empty space to the flight number.

Converted to draft so I can fix this.

**Comment by MisterChocker** [comment]
The flight number can be longer than that and the 3rd charactor from left does not "move" and stays put above the R.

**Comment by Username23-23** [comment]
> The flight number can be longer than that and the 3rd charactor from left does not "move" and stays put above the R.

Right, but I’m kind of confused why you think it’s the 3rd character from the left that doesn’t move rather than the 3rd from the right.
Looking at the video (references section), this was my thought process:

- The flight number in the video is EIN605E, where 05E is aligned with REC
- In the references, it is NW504 (2 characters shorter), where 504 is aligned with REC
- In both, it is the last 3 letters aligned with REC
- This also matches with the flight phase, because it’s two characters farther left in the video

All of this rests on the assumption that they both have the same spacing format. Feel free to correct me on anything.
Thanks for taking the time to review in this much detail!

**Comment by MisterChocker** [comment]
Oh I see, nvm then.

**Comment by Username23-23** [comment]
Alright, with the spacing issue, it looks like it’s going to take me a while. Right now, I’m planning on adding a bunch of spaces to the end of the line, and padStarting the flight phase with variable with spaces a number of times depending on the length of the flight number string. The logic for factoring in the length is really what’s taking a long time for me to come up with. If this is problematic or there’s a better way to do it that I’m overlooking, please let me know.

**Comment by MisterChocker** [comment]
I've looked into our FCOM, unfortunately this raises more questions than it gives answers. It shows a completely different spacing on the CRZ OPT REC MAX line and flight phase text in various different positions from far left to far right. (same for flight numbers). So I'd say we stick with what we got from the honeywell step 1 manual.

**Comment by Username23-23** [comment]
So I’ve been working on spacing, and this is what I have so far:
mail
As you can see, it’s more or less in the correct spot, but I haven’t been able to align it perfectly. It’s slightly off, just like in the before picture. I’ll try to fix this, but I’m not sure I’ll be able to.
Anyways, I’ll refine it, fix some issues, and then push around Saturday. Ignore the white text for the flight phase (it’s just for testing).

**Comment by MisterChocker** [comment]
This is how the takeoff page solved this issue. It's always the same length, and uses spaces `\xa0`.
["TAKE OFF RWY\xa0{green}" + runway.padStart(3, "\xa0") + "{end}[color]" + titleColor],

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 29.3.2021
Version of the sim: 1.14.6.0
PR Tested: #4959
Tier of testing: 2
Changes to observe: PROG page title line
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP


2.) After start up and on take off


3.) Climb


4.) Cruise


5.) Descent


6.) Land


7.) After land


Issues: -

Overall Rating: Good

Conclusions of the testing: Works

Hope this helps :-)

**Comment by MisterChocker** [comment]
The APPR is not aligned correctly.

**Comment by Username23-23** [comment]
> Hmm, the result does not match the 

How exactly? In the QA screenshots, the flight number is one character shorter than the one in the step manual, so the actual flight phase is aligned 1 space to the right of where it is in the manual i.e. where the manual example aligns the t in opt with the t in TO, the t in to should align 1 letter right in the QA screenshots (if that’s what you were talking about). Let me know if something else seems off.

**Comment by MisterChocker** [comment]
@Username23-23
1. The overall text is not aligned perfectly. (title vs line 1)(may need an additional space in the front)
2. The flight phase text positions don't match the reference.

**Comment by Username23-23** [comment]
> @Username23-23
> 
> 1. The overall text is not aligned perfectly. (title vs line 1)(may need an additional space in the front)
> 2. The flight phase text positions don't match the reference.

Could we discuss this on discord?

**Comment by MisterChocker** [comment]
Yes exactly!
Also we might need to add a leading space to perfect the text alignment with the other rows

**Comment by Username23-23** [comment]
> Yes exactly!
> Also we might need to add a leading space to perfect the text alignment with the other rows

I've fixed the spacing, but unfortunately nothing I've tried works for the alignment. Adding a leading space just makes the alignment issues even worse.

**Comment by MisterChocker** [comment]
@Username23-23 I'll take a look at it. Lets see what I can do.

**Comment by MisterChocker** [comment]
I've fixed the alignment issue by
- removing the central alignment of opt fl text,
- removing the central alignment of the opf fl,
- fixing the padding of the flight phase and flight number,
- by tricking the central alignment of the title.

I've used padStart and padEnd to ensure we always position the text correctly and maintaining the same amount of letters in the row.

Here the result:
image
image
image
image
image
image
image

**Comment by Username23-23** [comment]
> I've fixed the alignment issue by
> 
> * removing the central alignment of opt fl text,
> * removing the central alignment of the opf fl,
> * fixing the padding of the flight phase and flight number,
> * by tricking the central alignment of the title.
> 
> I've used padStart and padEnd to ensure we always position the text correctly and maintaining the same amount of letters in the row.
> 
> Here the result (spoiler)
> image
> image
> image
> image
> image
> image
> image
> 
> It maintains the exact same pos
> image
> image

I see. So the center alignment on the opt fl and the line below was the main issue?

**Comment by MisterChocker** [comment]
No, I fixed that alignment too, but the main issue was that the title was not aligned correctly.
In addition I fixed the overall padding.

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 31/3/21
Version of the sim: 1.14.6.0
PR Tested: #3958 
Tier of testing: 2
Changes to observe:
MCDU Progress page

Testing technique: Completed a full flight
Steps you've made to test the Pull Request:
1. Loaded in
2. looked at progress page after loading flight plan
3. looked at progress page in climb
4. in cruise
5. descent
6. and approach

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.
Taxi out
image
Climb
image
Cruise
image
Descent
image
Approach
image
Go Around
image
After landing
image


Issues (If Any): None that I can notice but not many references for the others so not fully sure if they are supposed to look like that
Steps to achieve the issue

Overall Rating: Good
Conclusions of the testing: LGTM, not sure if its all accurate as the references on the pr only shows the TO but looks works as expected.

**Comment by MisterChocker** [comment]
Please fix conflicts.

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 5.04.2021
Version of the sim: 1.14.6.0
PR Tested: #3959
Tier of testing: 2
Changes to observe: PROG page title line
Testing technique: Tryout

Steps made to test the Pull Request:

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) After start up (Image1) and on take off
3.) Climb
4.) Cruise
5.) Descent
6.) Land
7.) After land

Media:


Image1

Issues: OPT value is not shown. Goes from --- to blank.

Overall Rating: -

Conclusions of the testing: Small bug found. Rest works fine and looks good.

Hope this helps :-)

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js:8]:
```suggestion
        const flOpt = (mcdu._zeroFuelWeightZFWCGEntered && mcdu._blockFuelEntered && (mcdu.isAllEngineOn() || Simplane.getIsGrounded())) ? "{green}FL" + (Math.floor(flMax / 5) * 5).toString() + "{end}" : "-----";
```

**Comment by MisterChocker** [comment]
Good catch! Also the REC MAX magenta was do no longer magenta.
I've fixed it now.

image

**Pull request merged**
